### PR TITLE
Wheels

### PR DESCRIPTION
**Pull request by henryiii**
* Move testing to GHA
* Add wheel builds
* ~~Enable parallel compile~~ (seems to break, can be reenabled later)
* Use PEP 517/PEP 518 builds
* Simplify setup.py quite a bit
    - Use setuptools built-in Cython support
* Add `python_requires`

### CODE REVIEW

**Comment by kratsg** [comment]
Hi Henry, this looks great. Two questions:

1) what happens with copyrights? (I see you have your copyrights around a few of these files, but I'm not sure what the appropriate policy here is, given scikit-hep/iris-hep/etc...)
2) do we _really_ want to support python2?

**Comment by henryiii** [comment]
> 1. what happens with copyrights?

I think the general procedure is to put the author as the copyright holder and use a free license. It was discussed a few times and I think this was the conclusion.

> 2. do we _really_ want to support python2?

See <https://scikit-hep.org/supported-python-versions>. And all packages *must* have a release with python_requires before dropping Python 2 (if they ever supported Python 2, that is)!

**Comment by henryiii** [comment]
Passes!

**Comment by henryiii** [comment]
Standard reminder: needs to be regular-merged, as rebase breaks subtree (at least locally it does).

**Comment by kratsg** [comment]
> Standard reminder: needs to be regular-merged, as rebase breaks subtree (at least locally it does).

Will squash-n-merge work?

**Comment by henryiii** [comment]
Not to my knowledge. The “merge” commit must be preserved for it to be updatable later via a subtree pull. That’s why rebase fails, it thinks the merge commit is empty and drops it, which causes all the files to move out of their prefix and into the base repository! You have to rebase with mergers preserved (don’t know what github does for PRs). I’m by no means an expert, I prefer sub modules, but azure doesn’t check out submodules before configuring.

**Comment by kratsg** [comment]
If I look at squash+merge on GH, this is what I see

```
* Squashed '.ci/azure-wheel-helpers/' content from commit e52961a

git-subtree-dir: .ci/azure-wheel-helpers
git-subtree-split: e52961a9467954d8393b8fcbc714d44c97ec2166

* Add build, parallel, simplify
```

I'm generally a fan of a single commit, rather than N+merge -- but if this is the issue with subtree, we can do a regular merge here.

File: .ci/azure-pipeline-build.yml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+- job: macOS
+  variables:
+    python.architecture: 'x64'
+  strategy:
+    matrix:
+      Python27:
+        python.version: '2.7'
+      Python36:
+        python.version: '3.6'
```
Comment (eduardo-rodrigues) [inline .ci/azure-pipeline-build.yml]: No 3.5 here?

File: .ci/azure-wheel-helpers/README.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
+## Azure Wheel Helpers
+
+This repository holds a collection of wheel helpers designed by the [Scikit-HEP][] project to build Python Wheels on [Azure DevOps][]. This is designed for packages that require building; if you have a pure-Python project, producing a universal wheel is trivial without this helper collection. This collection assumes some standard paths and procedures, though *some* of them can be customized.
+
+Azure provides manual pipeline triggering and release pipelines, making it slighly better suited for this than GitHub Actions, though otherwise they are *very* similar.
+
+### Supported platforms and caveats
+
+TLDR: Python 2.7, 3.6, 3.7, and 3.8  on all platforms, along with 3.5 on Linux.
```
Comment (eduardo-rodrigues) [inline .ci/azure-wheel-helpers/README.md]: Should we just drop 3.5 altogether?

**Comment by eduardo-rodrigues** [comment]
> what happens with copyrights?
> 
> 
> I think the general procedure is to put the author as the copyright holder and use a free license. It was discussed a few times and I think this was the conclusion. @HDembinski @eduardo-rodrigues

Seems often the most adequate think to do, e.g. as in scikit-learn. For sure the copyright cannot be DIANA/HEP or IRIS-HEP or similar, but rather a university, etc. Having the main author is best is many situations.

File: .ci/azure-wheel-helpers/README.md (hunk 2 above)
Comment (henryiii) [reply .ci/azure-wheel-helpers/README.md]: We probably should soon, but I'd have to filter it out of the manylinux versions (like I filter 3.4 already). I'm hoping to later add a nice way to do this per project.

File: .ci/azure-pipeline-build.yml (hunk 1 above)
Comment (henryiii) [reply .ci/azure-pipeline-build.yml]: Windows and macOS tend to not be stuck on old versions like Linux users with system Pythons, and macOS doesn't have a pure 64bit 3.5 release I can easily use.

File: .ci/azure-wheel-helpers/README.md (hunk 2 above)
Comment (henryiii) [reply .ci/azure-wheel-helpers/README.md]: Anything in azure-wheel-helpers should not be changed, and can be updated from the main repository with a `git subtree pull <options>`.

**Comment by henryiii** [comment]
> I'm generally a fan of a single commit, rather than N+merge -- but if this is the issue with subtree, we can do a regular merge here.

I'd have to play with it to see what happens if it's squashed. I've not even done a subtree pull yet.

**Comment by henryiii** [comment]
> I'm generally a fan of a single commit, rather than N+merge -- but if this is the issue with subtree, we can do a regular merge here.

Github won't rebase-and-merge, it detects that the files will change if it does. And squash and merge does lose the history:

```
~/g/s/numpythia $ git subtree pull --prefix .ci/azure-wheel-helpers git@github.com:scikit-hep/azure-wheel-helpers.git master --squash
From github.com:scikit-hep/azure-wheel-helpers
 * branch            master     -> FETCH_HEAD
Can't squash-merge: '.ci/azure-wheel-helpers' was never added.

~/g/s/numpythia $ git checkout henryiii-wheels 
Switched to branch 'henryiii-wheels'
Your branch is up to date with 'origin/henryiii-wheels'.

~/g/s/numpythia $ git subtree pull --prefix .ci/azure-wheel-helpers git@github.com:scikit-hep/azure-wheel-helpers.git master --squash
From github.com:scikit-hep/azure-wheel-helpers
 * branch            master     -> FETCH_HEAD
Subtree is already at commit 02274e749eaeb9903b185c015b0d2f6d633a2913.
```

**Pull request merged**